### ISSUE TITLE
binutils tools should be part of Spack compiler definition

### ISSUE THREAD

**Issue by jonglezb** [LLNL/spack#21284]
binutils is currently taken from the system by default, even when using a compiler that lives under Spack. This is a bit strange given how important it is when building.

### Rationale

To make sure we don't depend on a particular system package, we bootstrap our Spack tree by building GCC within Spack, and we then use this Spack GCC as Spack compiler. Then, subsequent updates of our Spack tree are done using a different system that doesn't even have a system compiler.

The problem is that Spack keeps trying to use the system binutils, which completely fails when the system doesn't have a system compiler (GCC cannot build anything without `as`). We solve it in the following way:

- build binutils when bootstraping
- when updating the tree, start by loading binutils (`spack load binutils`)

However, it does not feel like this is how Spack should work.

### Description

According to Spack documentation, one way would be for all C packages to depend on binutils, but this is considered too annoying to do (I agree): https://spack.readthedocs.io/en/latest/getting_started.html?highlight=binutils#package-utilities

We could workaround by adding binutils to `prepend_path` in the compiler definition (I haven't tried that yet):
https://spack.readthedocs.io/en/latest/basic_usage.html?highlight=binutils#compiler-environment-variables-and-additional-rpaths
However, it feels more like a hack, and wouldn't work when auto-detecting compilers.

The cleanest solution would probably be to declare binutils tools in the compiler definition, e.g.:

```
- compiler:
    spec: gcc@8.3.0
    paths:
      cc: /grid5000/spack/opt/spack/linux-debian10-x86_64/gcc-8.3.0/gcc-8.3.0-cva45sgdtwukvazpxc6eurpry7xpi4zl/bin/gcc
      cxx: /grid5000/spack/opt/spack/linux-debian10-x86_64/gcc-8.3.0/gcc-8.3.0-cva45sgdtwukvazpxc6eurpry7xpi4zl/bin/g++
      f77: /grid5000/spack/opt/spack/linux-debian10-x86_64/gcc-8.3.0/gcc-8.3.0-cva45sgdtwukvazpxc6eurpry7xpi4zl/bin/gfortran
      fc: /grid5000/spack/opt/spack/linux-debian10-x86_64/gcc-8.3.0/gcc-8.3.0-cva45sgdtwukvazpxc6eurpry7xpi4zl/bin/gfortran
      as: /grid5000/spack/opt/spack/linux-debian10-x86_64/gcc-8.3.0/binutils-2.32-xiigusjo2rxs4yt4kuygdk3nf3qmuplm/bin/as
      ld: /grid5000/spack/opt/spack/linux-debian10-x86_64/gcc-8.3.0/binutils-2.32-xiigusjo2rxs4yt4kuygdk3nf3qmuplm/bin/ld
```

However, it might be very verbose if we specify all tools.

I'm open to other suggestions.  Maybe our usage is just not aligned with the way Spack works.

### Additional information

```
$ spack --version
0.16.0-908-85ea52acd1
```

### General information

- [x] I have run `spack --version` and reported the version of Spack
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by jonglezb** [comment]
So, I tried the `prepend_path` option and it works:

```
- compiler:
    spec: gcc@10.2.0
    paths:
      cc: /grid5000/spack/opt/spack/linux-debian10-ppc64le/gcc-8.3.0/gcc-10.2.0-tej2wa5tia6j3ec5pzsi5jkjs7eeebsg/bin/gcc
      cxx: /grid5000/spack/opt/spack/linux-debian10-ppc64le/gcc-8.3.0/gcc-10.2.0-tej2wa5tia6j3ec5pzsi5jkjs7eeebsg/bin/g++
      f77: /grid5000/spack/opt/spack/linux-debian10-ppc64le/gcc-8.3.0/gcc-10.2.0-tej2wa5tia6j3ec5pzsi5jkjs7eeebsg/bin/gfortran
      fc: /grid5000/spack/opt/spack/linux-debian10-ppc64le/gcc-8.3.0/gcc-10.2.0-tej2wa5tia6j3ec5pzsi5jkjs7eeebsg/bin/gfortran
    flags: {}
    operating_system: debian10
    target: ppc64le
    modules: []
    environment:
      prepend_path:
        PATH: /grid5000/spack/opt/spack/linux-debian10-ppc64le/gcc-8.3.0/binutils-2.35.1-vp3afr5tirb7k26hguaspp3jujrjaeag/bin
    extra_rpaths: []
```

If this is the preferred solution, I think it's more of a documentation issue. It should be made explicit instead of implicit.

More specifically, it should probably be mentioned in this page: https://spack.readthedocs.io/en/latest/getting_started.html#manual-compiler-configuration

Currently it's only described "on the side" in the "Specs & dependencies" section: https://spack.readthedocs.io/en/latest/basic_usage.html#compiler-environment-variables-and-additional-rpaths

**Comment by sethrj** [comment]
@jonglezb you did install `gcc+binutils`, correct? In my experience this does correctly set up the GCC compiler internals (with at least [one exception](https://github.com/spack/spack/issues/21870)) without the need for any shenanigans with `PATH`.

**Comment by benjaminjeliot** [comment]
@jonglezb I hit a problem where I was building ParaView with GCC, but one of the packages (ISPC) has a dependency on LLVM and uses the clang compiler instead. I think that GCC was using the correct binutils but clang wasn't, which caused errors.

Your `prepend_path` suggestion in https://github.com/spack/spack/issues/21284#issuecomment-773533721 seems to fix this issue.

Thanks :-)

**Comment by alalazo** [comment]
Fixed in v0.22 by https://github.com/spack/spack/pull/43956
